File: python/array/q48.py
def rotate(self, matrix):
    """
    :type matrix: List[List[int]]
    :rtype: void Do not return anything, modify matrix in-place instead.
    """
    for i in range(0,len(matrix)):
        for j in range(i+1,len(matrix)):
            matrix[i][j],matrix[j][i] = matrix[j][i],matrix[i][j]

    for i in range(0,len(matrix)):
        matrix[i].reverse()

        
void anti_rotate(vector<vector<int>>& matrix) {
    //take transpose and reverse reverse up-side-down
    for(int i = 0; i < matrix.size(); i++){
        for(int j = 0; j < matrix.size(); j++)
            swap(matrix[i][j],matrix[j][i]);
    }

    reverse(matrix.begin(),matrix.end());
}

File: c++/array/q48.cpp
void rotate2(vector<vector<int>>& matrix) {
    //Method 2: take the transpose and reverse left and right
    for(int i = 0; i < matrix.size(); i++){
        for(int j = i + 1; j < matrix.size(); j++)
            swap(matrix[i][j],matrix[j][i]);
    }

    for(int i = 0; i < matrix.size(); i++)
        reverse(matrix[i].begin(),matrix[i].end());
}

void rotate1(vector<vector<int>>& matrix) {
    //Method 1: reverse up-side-down, swap the symmetry
    reverse(matrix.begin(),matrix.end());

    for(int i = 0; i < matrix.size(); i++){
        for(int j = i + 1; j < matrix.size(); j++)
            swap(matrix[i][j],matrix[j][i]);
    }
}

void anti_rotate(vector<vector<int>>& matrix) {
    //take transpose and reverse reverse up-side-down
    for(int i = 0; i < matrix.size(); i++){
        for(int j = 0; j < matrix.size(); j++)
            swap(matrix[i][j],matrix[j][i]);
    }

    reverse(matrix.begin(),matrix.end());
}
